File: LeetCode-JavaScript/LeetCode 72.js
// LeetCode 72, 字符串的相似算法
// Levenshtein Distance

var minDistance = function(word1, word2) {
    let m = word1.length
    let n = word2.length
	console.log(`m: ${m} n: ${n}`)

    let dp = new Array();             //声明一维数组        
    for(var x = 0;x <= m;x++){
        dp[x] = new Array();         //声明二维数组
        for(var y=0;y <= n;y++){
            dp[x][y]=0;              //数组初始化为0
        }
    }

    for (let i = 0; i <= n; i++) {
        dp[0][i] = i
    }
    for (let i = 0; i <= m; i++) {
        dp[i][0] = i
    }

    for (let i = 1; i <= m; i++) {
        for (let j = 1; j <= n; j++) {
            if (word1[i] == word2[j]) {
                dp[i][j] = dp[i - 1][j - 1]
            } else {
                dp[i][j] = Math.min(
                    dp[i - 1][j] + 1,       // Insertions
                    dp[i][j - 1] + 1,       // Deletions
                    dp[i - 1][j - 1] + 1,   // Substitution
                )
            }
        }
    }
    return dp[m][n]
};

File: LeetCode-Java/src/DynamicProgramming/leetcode/LeetCode72.java
package DynamicProgramming.leetcode;

public class LeetCode72 {
    private String word1;
    private String word2;
    Integer[][] visited;
    public int minDistance(String word1, String word2) {
        visited = new Integer[word1.length()][word2.length()];
        return dp(word1.length() - 1, word2.length() - 1);
    }

    private int dp(int word1Index, int word2Index) {
        if (word1Index == -1) {
            /// word1 结束，word2 还剩 Word2Index + 1 个字符
            return word2Index + 1;
        }
        if (word2Index == -1) {
            return word1Index + 1;
        }
        if (visited[word1Index][word2Index] != null) {
            return visited[word1Index][word2Index];
        }
        if (word1.charAt(word1Index) == word2.charAt(word2Index)) {
            return dp(word1Index - 1, word2Index - 1);
        }
        // word1 在 word1Index 位置插入一个元素，此时 word2Index 前移，word1Index 不动；操作数 + 1
        int insert = dp(word1Index, word2Index - 1) + 1;
        int delete = dp(word1Index - 1, word2Index) + 1;
        int replace = dp(word1Index - 1, word2Index - 1) + 1;
        int res = Math.min(Math.min(insert, delete), replace);
        visited[word1Index][word2Index] = res;
        return res;
    }
}
